src/template/transaction/executor.ts

```diff
@@ -0,0 +1,822 @@
+  ins: z.infer<typeof ShuffleInstruction>
+) {
+  ctx.log(`Begin ShuffleInstruction: ${JSONBigInt.stringify(ins)}`);
+  const { target } = ins;
+
+  // The token array should never be shuffled since outputs have a "pointer" to the token position
+  // on the token array, so shuffling would make the outputs target different outputs.
+
+  if (target === 'inputs' || target === 'all') {
+    ctx.inputs = shuffle(ctx.inputs);
+  }
+
+  if (target === 'outputs' || target === 'all') {
+    ctx.outputs = shuffle(ctx.outputs);
+  }
+}
+
+/**
+ * Execution for ChangeInstruction
+ */
+export async function execChangeInstruction(
+  interpreter: ITxTemplateInterpreter,
+  ctx: TxTemplateContext,
+  ins: z.infer<typeof ChangeInstruction>
+) {
+  ctx.log(`Begin ChangeInstruction: ${JSONBigInt.stringify(ins)}`);
+  const token = getVariable<string | undefined>(ins.token, ctx.vars, ChangeInstruction.shape.token);
+  const address =
+    getVariable<string | undefined>(ins.address, ctx.vars, ChangeInstruction.shape.address) ??
+    (await interpreter.getChangeAddress(ctx));
+  const timelock = getVariable<number | undefined>(
+    ins.timelock,
+    ctx.vars,
+    ChangeInstruction.shape.timelock
+  );
+  ctx.log(`address(${address}) timelock(${timelock}) token(${token})`);
+
+  const tokensToCheck: string[] = [];
+  if (token) {
+    tokensToCheck.push(token);
+  } else {
+    // Check HTR and all tokens on the transaction
+    tokensToCheck.push(NATIVE_TOKEN_UID);
+    for (const tk of ctx.tokens) {
+      tokensToCheck.push(tk);
+    }
+  }
+
+  const script = createOutputScriptFromAddress(address, interpreter.getNetwork());
+
+  for (const tokenUid of tokensToCheck) {
+    ctx.log(`Checking change tx for token ${tokenUid}`);
+    const balance = ctx.balance.getTokenBalance(tokenUid);
+    const tokenData = ctx.addToken(tokenUid);
+    if (balance.tokens > 0) {
+      const value = balance.tokens;
+      ctx.log(`Adding a change output for ${value} / ${tokenUid}`);
+      // Need to create a token output
+      // Add balance to the ctx.balance
+      ctx.balance.addOutput(value, tokenUid);
+
+      // Creates an output with the value of the outstanding balance
+      const output = new Output(value, script, { timelock, tokenData });
+      ctx.addOutput(-1, output);
+    }
+
+    if (balance.mint_authorities > 0) {
+      const count = balance.mint_authorities;
+      ctx.log(`Adding ${count} mint authority change outputs / ${tokenUid}`);
+      // Need to create a token output
+      // Add balance to the ctx.balance
+      ctx.balance.addOutputAuthority(count, tokenUid, 'mint');
+
+      // Creates an output with the value of the outstanding balance
+      const output = new Output(TOKEN_MINT_MASK, script, {
+        timelock,
+        tokenData: tokenData | TOKEN_AUTHORITY_MASK,
+      });
+      ctx.addOutput(-1, ...Array(count).fill(output));
+    }
+
+    if (balance.melt_authorities > 0) {
+      const count = balance.mint_authorities;
```

Comment:
```suggestion
      const count = balance.melt_authorities;
```

Maybe we should add all possibilities to the tests.

Reply:
Done